balta/src/main/scala/za/co/absa/db/balta/classes/QueryResult.scala

```diff
@@ -16,32 +16,44 @@
 
 package za.co.absa.db.balta.classes
 
-import java.sql.ResultSet
+import za.co.absa.db.balta.classes.QueryResultRow.{Extractors, FieldNames}
+
+import java.sql.{ResultSet, ResultSetMetaData, SQLException}
 
 /**
  *  This is an iterator over the result of a query.
  *
  * @param resultSet - the JDBC result of a query
  */
 class QueryResult(resultSet: ResultSet) extends Iterator[QueryResultRow] {
-  private [this] var resultSetHasNext: Option[Boolean] = Some(resultSet.next())
+  val resultSetMetaData: ResultSetMetaData = resultSet.getMetaData
+  val columnCount: Int = resultSetMetaData.getColumnCount
+
+  private [this] var nextRow: Option[QueryResultRow] = None
+
+  private [this] implicit val fieldNames: FieldNames = QueryResultRow.fieldNamesFromMetdata(resultSetMetaData)
+  private [this] implicit val extractors: Extractors = QueryResultRow.createExtractors(resultSetMetaData)
 
   override def hasNext: Boolean = {
-    resultSetHasNext.getOrElse {
-      val result = resultSet.next()
-      resultSetHasNext = Some(result)
-      result
+    if (nextRow.isEmpty) {
+      try {
+        if (resultSet.next()) {
+          nextRow = Some(QueryResultRow(resultSet))
+        }
+      } catch {
+        case _: SQLException =>  // TODO Do nothing
```

Comment:
what are the possible causes of this? TODO indicates that you might want to discuss this further?

Reply:
Forgotten TODO. I was actually re-throwing the exception for debugging purposes. Removed the re-throw, forgot the TODO note.
Will remove.

Reply:
sure, I kind of thought so, it makes sense 👍